107-tee-boot-attestation.md

```diff
@@ -0,0 +1,497 @@
+For example, consider both security and how this will impact the larger
+enhancement ecosystem.
+
+How will security be reviewed and by whom?
+-->
+
+Keylime will store a TEE attestation key used to encrypt initial vTPM state for
+all encrypted guests attesting with the server. If this key is leaked, all of
+the images could potentially be decrypted and secrets can be leaked.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.  Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+-->
+
+ * New tests will need to be written specifically for TEE boot attestation
+   extension scenarios.
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this enhancement
+-->
+
+TODO
+
+### Dependency requirements
+
+<!--
+If your new change requires new dependencies, please outline and demonstrate that your selected dependency 
+is well maintained and packaged in Keylime's supported Operating Systems (currently Debian Stable
+and as of time writing Fedora 32/33). 
+
+During code implementation you will also be expected to add the package to CI , the keylime ansible role and 
+keylimes main installer (`keylime/installers.sh`).
+
+If the package is not available in the supported Operated systems, the PR will not be merged into master. 
+
+Adding the package in `requirements.txt` is not sufficent for master which is where we tag releases from. 
+
+You may however be able to work within an experimental branch until a package is made available. If this is
+the case, please outline it in this enhancement.
+
+-->
+No additional dependencies should be required.
+
+## Drawbacks
+
+<!--
+Why should this enhancement _not_ be implemented?
+-->
+No drawbacks are known of.
+
+## Alternatives
+
+<!--
+What other approaches did you consider and why did you rule them out?  These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+ * There are other TEE attestation servers released now, such as the [trustee]
```

Comment:
@fitzthum I mentioned a few concerns with using Trustee here. Would like to continue the Trustee discussion, PTAL at the process I outlined above and share your thoughts/where you see Trustee fitting.

Reply:
There are two potential ways that Trustee and Keylime can be combined. 

First, since Trustee handles confidential attestation but not runtime attestation and Keylime is the opposite, you can have Trustee do the initial confidential attestation and provision state for a vTPM which will be registered with Keylime. There is some coordination required here, but it is relatively simple. I think this is the approach you are referring to below. To me this is the best way to do confidential attestation. You mentioned that it would require two servers to run, but I don't see that as a problem. In fact to some people having two simpler services handling distinct things is preferable to combining everything. I don't think the number of processes running is what we should be optimizing for. Either way, this approach is almost certainly simpler and easier to implement than adding confidential attestation support to Keylime itself.

If you really want to tack on a blocking confidential attestation protocol to Keylime, then you should at least think about using the verifier crate provided by Trustee. This allows you to verify 8 different types of confidential evidence and avoid rolling your own security-sensitive code.

Reply:
> There are two potential ways that Trustee and Keylime can be combined.
> 
> First, since Trustee handles confidential attestation but not runtime attestation and Keylime is the opposite, you can have Trustee do the initial confidential attestation and provision state for a vTPM which will be registered with Keylime. There is some coordination required here, but it is relatively simple. I think this is the approach you are referring to below. To me this is the best way to do confidential attestation. You mentioned that it would require two servers to run, but I don't see that as a problem. In fact to some people having two simpler services handling distinct things is preferable to combining everything.

This is true if they were handling two distinct things, but we merge some of these pieces. For instance, trustee wouldn't have to simply provide initial vTPM state, but also extend a PCR with some value provided by the keylime verifier to indicate a successful attestation when the keylime agent registers itself post-boot. This requires much more coordination between the two servers and can get messy very fast.

> I don't think the number of processes running is what we should be optimizing for. Either way, this approach is almost certainly simpler and easier to implement than adding confidential attestation support to Keylime itself.

I disagree. Outside of providing TEE verifiers, this is really only making some slight changes to the registrar to encrypt/decrypt secrets and write extra information to the registrar DB.

> 
> If you really want to tack on a blocking confidential attestation protocol to Keylime, then you should at least think about using the verifier crate provided by Trustee. This allows you to verify 8 different types of confidential evidence and avoid rolling your own security-sensitive code.

I'm **much** more open to this approach as by using the verifier crate in trustee, we can simply provide a web hook to the verifier to attest evidence. As you also mention, we don't have to roll our own TEE verifier code.

For example with SEV-SNP, a webhook could be exposed to the registrar that would take the attestation report and certificates, verifying and giving a pass/fail.

Reply:
> This is true if they were handling two distinct things, but we merge some of these pieces. For instance, trustee wouldn't have to simply provide initial vTPM state, but also extend a PCR with some value provided by the keylime verifier to indicate a successful attestation when the keylime agent registers itself post-boot. This requires much more coordination between the two servers and can get messy very fast.

I think confidential attestation and runtime attestation are pretty distinct. I don't think Trustee needs to worry about the nonce during registration time. It can simply deliver a vTPM with a known identity and then the keylime agent can extend the PCRs as needed (i.e with a nonce). The presence of particular root keys in the vTPM will imply that a hardware attestation has taken place without Keylime having to know anything about the details of the confidential TCB. Keylime can use the vTPM like it would any other, which seems convenient.